drivers/at/at.c

```diff
@@ -258,3 +258,65 @@ ssize_t at_readline(at_dev_t *dev, char *resp_buf, size_t len, bool keep_eol, ui
     }
     return res;
 }
+
+#ifdef MODULE_AT_OOB
+void at_add_oob(at_dev_t *dev, at_oob_t *oob)
+{
+    assert(oob);
+    assert(oob->urc);
+    assert(oob->cb);
+
+    clist_rpush(&dev->oob_list, &oob->list_node);
+}
+
+void at_remove_oob(at_dev_t *dev, at_oob_t *oob)
+{
+    clist_remove(&dev->oob_list, &oob->list_node);
+}
+
+static int _check_oob(clist_node_t *node, void *arg)
+{
+    const char *buf = arg;
+    at_oob_t *oob = container_of(node, at_oob_t, list_node);
+
+    DEBUG("Trying to match with %s\n", oob->urc);
+
+    if (strncmp(buf, oob->urc, strlen(oob->urc)) == 0) {
+        oob->cb(oob->arg, buf);
+        return 1;
+    }
+
+    return 0;
+}
+
+void at_process_oob(at_dev_t *dev)
+{
+    const char eol[] = AT_RECV_EOL_1 AT_RECV_EOL_2;
+    assert(sizeof(eol) > 1);
+    char buf[AT_BUF_SIZE];
+    char *p;
+    size_t read = 0;
+
+    while (1) {
+        int res = isrpipe_try_read(&dev->isrpipe, buf + read, sizeof(buf) - read);
```

Comment:
I'm uncertain if it's good method to bypass the `isrpipe mutex`. I know an URC can happen at any time, but the `uart_init` callback to `character received` is connected to `isrpipe_write_one` which unlocks the `isrpipe mutex`. 

The way I used this method:
A `gpio_int` connected to the modem ring indicator. This signals a thread wakeup. Thread claims modem mutex and calls `at_process_oob`. And URC is dispatched to the correct callbacks.

If an URC is sent at any time, all data should be filtered for `+urc:data EOL` if the URC exists as callback, it should be stored and passed when `at_process_oob` is called. Otherwise the URC is discarded by active operations

![image](https://user-images.githubusercontent.com/37693428/41221526-a60f2474-6d64-11e8-920e-a41d2cd22956.png)

Reply:
AFAIK, an URC cannot come at anytime (i.e. not during the process of a command).
I'm also not sure to fully understand your point, and I have the impression that your proposal would mean having 2 isrpipes.

@toonst will continue working on this topic on our side, so I guess you can keep this discussion with him ;)

Reply:
I see U-Blox has a well documented AT-manual. U-Blox states "module delayes URC during command execution." Quectel and Telit don't specify this.... 

Example:
An `AT+COPS=?` command can last for 180s, a SMS can be received in this period of time. Or an incomming call present itself.

I'm not saying it should be changed, I think it can be fixed if a problem occurs.

Reply:
I don't really understand the issue with 'bypassing' the isrpipe mutex. As far as I understand, this mutex is only there for synchronization and to block the caller thread if the pipe is empty. 

In our case, we don't want to block if empty, so no need to wait for the mutex to be unlocked.

But as I read the isrpipe/tsrb code, I have troubles to understand how the tsrb is actually thread-safe...


To solve the 'an URC can come at any time' issue, we should process URC each time a command is sent and do as you suggested before I guess. But this make everything really complicated.

Reply:
I have a use case where the isrpipe should be locked:
I've left the `isrpipe` out of the drawing, but let the `isrpipe->mutex` in.

![image](https://user-images.githubusercontent.com/37693428/41296563-7f5ba7a4-6e5d-11e8-95bb-1cdbb06503d2.png)

Reply:
I'm not sure to have understand everything, but `isrpipe_read` first tries to read from the buffer, then lock the mutex if it needs to wait for more characters.

Then if your GSM thread has a higher priority than Main thread, it will try to read from the buffer regardless of the mutex state. So there are 2 cases:
- you can read everything right away, so you check your URC and return
- you cannot read everything and you lock the mutex, then the main thread will be scheduled first because it has locked the mutex before, so the URC (or a part of it) will go to the main thread

I have the feeling that we're trying to solve a problem without knowing if it really exists. I would start by locking the whole `at_dev_t` with a `rmutex` and consider that an URC cannot be received in the middle of a command.

Reply:
@vincent-d let me try over (I will skip the URC in the middle of a command/request):

I think the `at_process_urc` doesn't work because `isrpipe_try_read` only returns the number of read bytes. The process is expecting a -1 if `tsrb` is dried out, but it doesn't return this option.
```
int res = isrpipe_try_read(&dev->isrpipe, buf + read,  sizeof(buf) - read);
if (!res) {
    return;
}
```

Also I'm using the ring functionality of the modem. This notifies the host something is incoming. So my thread can sleep and wake on a notify. Instead of running the 'at_process_urc` endless (as you did in your modem pr).

Between a ring indication and an URC is usually  a few milliseconds, I would like a block functionality int the `at_process_urc` to block until a URC is received. e.g. `void at_process_oob(at_dev_t *dev, uint32_t timeout)`.
If timeout is 0 it doesn't block, if timeout is set it's blocked until a URC is completed (with EOL).

I can make this proposal if you like.

Reply:
Alright, I understand, finally ^^. I'll try to come up with something, you can also make a proposal if you want.

But still, I'm wondering if a lock at the `at_dev_t` level wouldn't make sense too.